angr/analyses/cfg/cfg_base.py

```diff
@@ -1719,6 +1723,46 @@ def _addr_to_function(self, addr, blockaddr_to_function, known_functions):
 
         return f
 
+    def _is_tail_call_optimization(self, src_addr, dst_addr, src_function, all_edges, known_functions,
+                                       blockaddr_to_function):
+        """
+        If source and destination belong to the same function, and the following criteria apply:
+        - source node has only one default exit
+        - destination is in front of the function that the source node belongs to
```

Comment:
If by "in front" you mean that the destination address should be greater than the source function address, then this condition doesn't always hold. For example, a function might end with a tail-call to one of the functions from *.plt* section, which is, to my knowledge, always located before the *.text* section.

However, from the code below I see that the actual condition is `dst_addr < src_function.addr`, and that condition doesn't always hold too. For example, the float-related functions that are commonly found in ARM binaries have all the kinds of weird jumps, including the tail-calls to functions, which are located below (target address is greater than the source address).

Reply:
Agree. It was actually my left-over debugging code.